Add sequential process

jrising opened the pull request:
This includes two changes needed to support the new labor clipping method (https://github.com/ClimateImpactLab/impact-calculations/pull/2).
 - `DelayedCurveGenerator` keeps track of the curve it most recently produced. This allows the clipping curve generator to get everything it needs from the dependent curve generator.
 - [Side-note: `DelayedCurveGenerator` actually always needed this, in case it was asked twice for a curve in a given year. The current implementation will give the next curve, for all requests after the first request.]
 - There's a new `SequentialProcess` function for calculations. This is now added implicitly by `create_calcstep` in `impact-calculations` in cases where a previous result is not used in a subsequent step.

Consider the following example:
```
calculation:
  - YearlyAverageDay:
      model: model1
  - YearlyAverageDay:
      model: model2
  - Rebase
```

Previously, only `model2` would ever be run, since its used by `Rebase`. But clearly the user intends for `model1` to be run (and presumably the results to be reported). So, upon reading the second `YearlyAverageDay` operation, both of these operations are wrapped in a `SequentialProcess`, and the resulting output (prior to the Rebase) is
```model2(T), model1(T)```.

jrising committed:
add sequential proc

jrising committed:
switch to cache

jrising committed:
put into curve cache instead

brews reviewed (commented):
Hmmm, interesting... I have some comments and feels but otherwise, I'd say this works and "LGTM".

* This new `DelayedCurveGenerator.get_most_recent_curve()` feels like unpythonic java/C++ OOP. Unless we have a good reason to use a "getter" then I'd just access `self.curve_cache` directly. If we *do* want to hide it then consider a `@property` and rename the new dictionary to `self._curve_cache`.
* Related: the name `self.curve_cache` and the distinction between it and `self.last_curves` is a bit hard to grok. Hmm. This is tricky. If you opt to access it directly, then consider renaming it to something like `self.current_curves`, `self.current_curves_cache` or `self.recent_curves`. 
* Do we now have reason to access `self.last_curve()` outside of `DelayedCurveGenerator.get_curve()`? If not, then maybe rename it to `self._last_curve()` to communicate this is special and private-ish. Nevermind if this would break things because it's used elsewhere.
* `SequentialProcess` and `SequentialProcessApplication` have no docstrings or tests for new behavior.

jrising commented on openest/generate/functions.py:
```diff
@@ -1148,3 +1148,61 @@ def describe():
+
+    def format(self, lang, *args, **kwargs):
+        beforeauxlen = len(formatting.format_labels)
+        auxres = self.step1.format(lang, *args, **kwargs)
```
`format` creates labels that can then be used by other `format` calls. This can be cleaned up and made more clear-- like so many changes, I had to add it in a rush way-back-when.

jrising committed:
add docs and test

jrising committed:
typo

jrising committed:
...

jrising committed:
typo

jrising commented:
Okay, I've removed `get_most_recent_curve` and changed the name of the attribute to a clearer `current_curves`. I've also added a docstring and a simple unit test.

`last_curve` is used by at least one child class, so I think it's a little unclear whether we want to rename it. I think we can let it ride for now.

jrising committed:
add extrema_finder

jrising commented:
Merging this in.
